sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,2792 @@
+// Copyright lowRISC contributors.
```

Comment:
> IMHO, just relying on CSR_READ(name, output), CSR_WRITE(name, value) should be enough and would be much leaner.

Personally I don't think the header file is so large/complex that it is a problem. I think the API it provides is more readable and easier to use correctly than the equivalent API using macros. By far most of the line count consists of documentation. We could document the functions in chunks if we want to reduce the size of the file. Also note that not all CSR names are supported by the assembler so we even with macros we will need to at least have some constants to help with using them.

> I also noticed that your implementation returns the old value of a CSR after a write. This could be useful but I couldn't find any call-sites that relies on this (did I miss something?). I was wondering maybe we should not return the old value to simplify the API and tests.

There are no uses in tree but I am currently using it in the ePMP implementation to reduce the number of CSR operations. In general I think it might be useful when hardening but I agree with you that removing it would make tests a bit simpler.

Reply:
About returning the old value: The reason that `csrr` does this is to make it simpler to write exception-safe code. It can be useful when you want to save and modify (and later restore) e.g. some configuration register without needing to worry about what happens if an instruction hits just between the read and write.

I'd expect this to be useful in e.g. the interrupt vector (if we're handling nested interrupts).

Reply:
I think we should keep the return values then. I think we can make testing easier by just making the return value optional in the `EXPECT_CSR_*` macros.

Reply:
> I also noticed that your implementation returns the old value of a CSR after a write. This could be useful but I couldn't find any call-sites that relies on this (did I miss something?). I was wondering maybe we should not return the old value to simplify the API and tests.

We should try to figure out a way to ensure the value must be used. Can we hit values with `warn_unused_result`?

I suggest adding a function like this and passing the return values through it:
```
__attribute__((warn_unused_result)) //
inline uint32_t csr_force_unused_warning(uint32_t x) { return x; }
```

Reply:
Unfortunately I don't think there is a way to get the compiler to warn about unused macro output parameters. If we ever switch back to generated functions in the future we should do this for `csr_read` though.